File: LC/python/Recover Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a tree node
    def recoverTree(self, root):
        self.n1 = self.n2 = None
        self.prev = None
        self.find(root)
        self.n1.val,self.n2.val = self.n2.val,self.n1.val
        return root
        
    def find(self,root):
        if root:
            self.find(root.left)
            if self.prev and root.val<self.prev.val:
                self.n2 = root
                if self.n1 == None:
                    self.n1 = self.prev
            self.prev = root
            self.find(root.right)
                

File: LC/java/Recover Binary Search Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode p1,p2,pre;
    
    public void recoverTree(TreeNode root) {
        find(root);
        int tmp;
        tmp = p1.val;
        p1.val = p2.val;
        p2.val = tmp;
    }
    
    public void find(TreeNode root){
        if(root != null){
            find(root.left);
            if (pre!=null && root.val<pre.val){
                p1 = root;
                if (p2 == null){
                p2 = pre;
                }
            }
            pre = root;
            find(root.right);
        }
    }
}